### PR TITLE
[GH-8327] Deprecate EntityManagerHelper for a provider abstraction.

### LINKED ISSUE #8327: cli-config.php compatibility with DBAL
Since DBAL 2.11.0 and https://github.com/doctrine/dbal/pull/3956, it is now deprecated to use HelperSet in `cli-config.php`. This will become impossible in DBAL 3, following https://github.com/doctrine/dbal/pull/4059.

This means that, in order to keep sharing a single config file for both packages, this package should also accept the new `\Doctrine\DBAL\Tools\Console\ConnectionProvider`. It could actually be a new interface along the lines of:

```php
namespace \Doctrine\ORM\Tools\Console;

interface EntityManagerProvider
    public function getEntityManager(string $name) : EntityManagerInterface;
}
```

I suppose the support of HelperSet could be drop entirely in the next major version too. So long we can use a `HelperSetManagerProvider` that extends the `EntityManagerProvider`.

All commands must be adjusted to access the entity manager from a provider instead of the HelperSet. Deprecation messages when using HelperSet must be thrown appropriately.

See related https://github.com/doctrine/migrations/issues/1070

### PR DESCRIPTION
**Pull request by beberlei**
Deprecates the `EntityManagerHelper` in favor of a new `EntityManagerProvider` interface that comes with two implementation.

- A BC abstraction wrapping the Console HelperSet `HelpersetManagerProvider`
- a new `SingleManagerProvider` that provides access to a single entity manager.

In addition an implementation for the DBAL `ConnectionProvider` is provided that is based on any `EntityManagerProvider` and calling `getConnection()` on the EntityManager.

It is expected for the DoctrineBundle to create a new `EntityManagerProvider` that provides access to their multiple ORM entity managers via the Symfony DI Container.

Two deprecation messages are added:
- When the `HelperSetManagerProvider` must be used from within `ConsoleRunner`, because a `HelperSet` is passed.
- When an ORM command is used without explicitly passing the `EntityManagerProvider` as a dependency.

Todos:
- [x] Sort out update of DBAL and 7.2 dependency, waiting for https://github.com/doctrine/dbal/pull/4386
- [x] Update tests to pass provider

Fixes #8327

### CODE REVIEW

File: lib/Doctrine/ORM/Tools/Console/EntityManagerProvider/SingleManagerProvider.php [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+    public function __construct(EntityManagerInterface $entityManager)
+    {
+        $this->entityManager = $entityManager;
+    }
+
+    public function getManager(string $name = 'default'): EntityManagerInterface
+    {
+        if ($name !== 'default') {
+            throw UnknownManagerException::unknownManager($name, ['default']);
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Tools/Console/EntityManagerProvider/SingleManagerProvider.php:21]: Is it possible that people define several managers but end up using `SingleManagerProvider`? If yes, that exception will be confusing I think.

File: tests/Doctrine/Tests/ORM/Tools/Console/ConsoleRunnerTest.php [hunk 2]
```diff
@@ -60,4 +66,12 @@ public function testCreateApplicationShouldAppendGivenCommands(): void
 
         self::assertTrue($app->has($command));
     }
+
+    public function testCreateApplicationWithProvider(): void
+    {
+        $provider = $this->getMockBuilder(EntityManagerProvider::class)->getMock();
```
Comment (greg0ire) [inline tests/Doctrine/Tests/ORM/Tools/Console/ConsoleRunnerTest.php]:
```suggestion
        $provider = $this->createMock(EntityManagerProvider::class);
```

File: lib/Doctrine/ORM/Tools/Console/EntityManagerProvider/SingleManagerProvider.php (hunk 1 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/Tools/Console/EntityManagerProvider/SingleManagerProvider.php:21]: Also, let's test this?

Comment (beberlei) [reply lib/Doctrine/ORM/Tools/Console/EntityManagerProvider/SingleManagerProvider.php:21]: No the `SingleManagerProvider` is for non Symfony context, where usually you have only one and you don't care about any of this. :)

File: lib/Doctrine/ORM/Tools/Console/Command/AbstractEntityManagerCommand.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    private $entityManagerProvider;
+
+    public function __construct(?EntityManagerProvider $entityManagerProvider = null)
+    {
+        parent::__construct();
+
+        $this->entityManagerProvider = $entityManagerProvider;
+    }
+
+    protected function getEntityManager(InputInterface $input): EntityManagerInterface
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Tools/Console/Command/AbstractEntityManagerCommand.php]:
```suggestion
    final protected function getEntityManager(InputInterface $input): EntityManagerInterface
```

File: lib/Doctrine/ORM/Tools/Console/EntityManagerProvider/ConnectionFromManagerProvider.php [hunk 4]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace Doctrine\ORM\Tools\Console\EntityManagerProvider;
+
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Tools\Console\ConnectionProvider;
+use Doctrine\ORM\Tools\Console\EntityManagerProvider;
+
+class ConnectionFromManagerProvider implements ConnectionProvider
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Tools/Console/EntityManagerProvider/ConnectionFromManagerProvider.php]:
```suggestion
final class ConnectionFromManagerProvider implements ConnectionProvider
```

File: lib/Doctrine/ORM/Tools/Console/EntityManagerProvider/HelperSetManagerProvider.php [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+use Doctrine\Deprecations\Deprecation;
+use Doctrine\ORM\EntityManagerInterface;
+use Doctrine\ORM\Tools\Console\EntityManagerProvider;
+use Doctrine\ORM\Tools\Console\Helper\EntityManagerHelper;
+use Symfony\Component\Console\Helper\HelperSet;
+
+use function assert;
+
+class HelperSetManagerProvider implements EntityManagerProvider
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Tools/Console/EntityManagerProvider/HelperSetManagerProvider.php]:
```suggestion
final class HelperSetManagerProvider implements EntityManagerProvider
```

File: lib/Doctrine/ORM/Tools/Console/EntityManagerProvider/SingleManagerProvider.php [hunk 6]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+namespace Doctrine\ORM\Tools\Console\EntityManagerProvider;
+
+use Doctrine\ORM\EntityManagerInterface;
+use Doctrine\ORM\Tools\Console\EntityManagerProvider;
+
+class SingleManagerProvider implements EntityManagerProvider
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Tools/Console/EntityManagerProvider/SingleManagerProvider.php]:
```suggestion
final class SingleManagerProvider implements EntityManagerProvider
```

File: lib/Doctrine/ORM/Tools/Console/EntityManagerProvider/UnknownManagerException.php [hunk 7]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+namespace Doctrine\ORM\Tools\Console\EntityManagerProvider;
+
+use OutOfBoundsException;
+
+use function implode;
+use function sprintf;
+
+class UnknownManagerException extends OutOfBoundsException
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Tools/Console/EntityManagerProvider/UnknownManagerException.php]:
```suggestion
final class UnknownManagerException extends OutOfBoundsException
```

**Review by greg0ire — DISMISSED** [review]
LGTM. The code suggestions I added are just that: suggestions

**Comment by dominikzogg** [comment]
@beberlei adding `em` option probably breaks every glue code out there similar to the doctrine bundle. This change should be in 3.0 not 2.9

**Comment by beberlei** [comment]
@dominikzogg we speifically added code to prevent breaking in doctrine Bundle

**Comment by beberlei** [comment]
https://github.com/doctrine/DoctrineBundle/commit/d6b3c37804539a24ba8a7d647a6144cab2f13242

**Comment by dominikzogg** [comment]
@beberlei as mentioned: there is code outside the symfony ecosystem out there

https://github.com/chubbyphp/chubbyphp-laminas-config-doctrine/pull/3/files and i am not even done yet => will lead to a major release on my side and when orm 3.0 gets released another major cause i am sure there will be (expected) bc's

**Comment by beberlei** [comment]
Adding *any* parameter could have caused this, others might use `entity-manager` instead of `em`. Not sure how this could have been prevented, we won't change this now. Sorry to cause extra work on your end.

**Comment by dominikzogg** [comment]
> Not sure how this could have been prevented

do it in 3.0, so nobody get forced to update code immediately

**Comment by beberlei** [comment]
That ship has sailed :)

**Pull request merged**